adcircpy/mesh/base.py

```diff
@@ -367,49 +260,118 @@ def __call__(self) -> gpd.GeoDataFrame:
-        return gpd.GeoDataFrame(data, crs=self._grd.crs)
+        return GeoDataFrame(data, crs=self._grd.crs)
 
-    @lru_cache(maxsize=1)
-    def implode(self) -> gpd.GeoDataFrame:
-        return gpd.GeoDataFrame(
-            {'geometry': MultiPolygon([polygon.geometry for polygon in self().itertuples()])},
+    @property
+    def implode(self) -> GeoDataFrame:
+        return GeoDataFrame(
+            {
+                'geometry': MultiPolygon(
+                    [polygon.geometry for polygon in self.geodataframe.itertuples()]
+                )
+            },
             crs=self._grd.crs,
         )
 
+    @property
     @lru_cache(maxsize=1)
     def multipolygon(self) -> MultiPolygon:
-        polygon_collection = []
-        for rings in self.rings.sorted().values():
-            exterior = self._grd.nodes.coord[rings['exterior'][:, 0], :]
-            interiors = []
-            for interior in rings['interiors']:
-                interiors.append(self._grd.nodes.coord[interior[:, 0], :])
-            polygon_collection.append(Polygon(exterior, interiors))
-        mp = MultiPolygon(polygon_collection)
-        if isinstance(mp, Polygon):
-            mp = MultiPolygon([mp])
-        return mp
+        triangles = self._grd.elements.triangulation.triangles
+
+        triangle_edges, counts = numpy.unique(
+            numpy.sort(
+                numpy.concatenate(
+                    [triangles[:, :2], triangles[:, 1:], triangles[:, [0, 2]]], axis=0
+                ),
+                axis=1,
+            ),
+            axis=0,
+            return_counts=True,
+        )
+        boundary_edges = triangle_edges[counts == 1]
+
+        boundary_edge_points = self._grd.nodes.iloc[:, :2].values[boundary_edges]
+
+        exterior_polygons = collect_interiors(list(polygonize(boundary_edge_points.tolist())))
+
+        coords = self._grd.nodes.values
+        x = coords[:, 0]
+        y = coords[:, 1]
+        total_triangle_area = numpy.sum(
+            numpy.abs(
+                (
+                    x[triangles[:, 0]] * (y[triangles[:, 1]] - y[triangles[:, 2]])
+                    + x[triangles[:, 1]] * (y[triangles[:, 2]] - y[triangles[:, 0]])
+                    + x[triangles[:, 2]] * (y[triangles[:, 0]] - y[triangles[:, 1]])
+                )
+                / 2
+            )
+        )
+
+        if exterior_polygons[-1].area < total_triangle_area:
+            polygon_collection = []
+            coords = self._grd.coords.values
+            for rings in self.sorted().values():
+                exterior = coords[rings['exterior'][:, 0], :]
+                interiors = []
+                for interior in rings['interiors']:
+                    interiors.append(coords[interior[:, 0], :])
+                polygon_collection.append(Polygon(exterior, interiors))
+
+            exterior_polygons.extend(polygon_collection)
+            exterior_polygons = collect_interiors(exterior_polygons)
```

Comment:
if initial `polygonize` method doesn't work, revert to previous algorithm for finding and sorting edges